### PR TITLE
Scope and collection API for template.

### LINKED ISSUE #963: Support for Scope and Collection API Changes [DATACOUCH-655]
**[Arun Vijayraghavan](https://jira.spring.io/secure/ViewProfile.jspa?name=JIRAUSER49896)** opened **[DATACOUCH-655](https://jira.spring.io/browse/DATACOUCH-655?redirect=false)** and commented

Not sure if this ticket is needed, but this will stand as a place holder for the API changes w.r.t underlying server specific scopes and collections API Changes.

 

Please feel free to close this ticket if you think this will be taken care by the JAVA SDK

---
No further details from [DATACOUCH-655](https://jira.spring.io/browse/DATACOUCH-655?redirect=false)

### PR DESCRIPTION
**Pull request by mikereiche**
Scope and collection API for template.

Closes #963.
Original pull request: #1071.

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: src/test/java/org/springframework/data/couchbase/repository/ReactiveCouchbaseRepositoryQueryIntegrationTests.java [hunk 1]
```diff
@@ -72,7 +76,7 @@ void shouldSaveAndFindAll() {
 		Airport vie = null;
 		Airport jfk = null;
 		try {
-			vie = new Airport("airports::vie", "vie", "loww");
+			vie = new Airport("airports::vie", "vie", "low1");
```
Comment (mikereiche) [inline src/test/java/org/springframework/data/couchbase/repository/ReactiveCouchbaseRepositoryQueryIntegrationTests.java:79]: After running tests, I found some "vie", "loww" Airport records. So I made them unique to figure out what test was creating them and not deleting them.

File: src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java [hunk 2]
```diff
@@ -59,7 +59,7 @@ ClusterType type() {
 
 	@Override
 	TestClusterConfig _start() throws Exception {
-		bucketname = UUID.randomUUID().toString();
+		bucketname = "my_bucket"; // UUID.randomUUID().toString();
```
Comment (mikereiche) [inline src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java:62]: Reuse "my_bucket" if it already exists.  Creating the indexes, scopes and collections takes time.

**Review by daschl — CHANGES_REQUESTED** [review]
This change looks veeery solid already, most of  my comments are nits and small changes -- should be able to merge afterwards

File: src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplate.java [hunk 3]
```diff
@@ -184,4 +185,12 @@ private void prepareIndexCreator(final ApplicationContext context) {
 	TemplateSupport support() {
 		return templateSupport;
 	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public void setThreadLocalArgs(PseudoArgs pseudoArgs) {
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplate.java:192]: It's never used.  Removed.

File: src/main/java/org/springframework/data/couchbase/core/ExecutableReplaceByIdOperation.java [hunk 4]
```diff
@@ -18,46 +18,117 @@
...
-import org.springframework.data.couchbase.core.support.WithCollection;
+import org.springframework.data.couchbase.core.support.WithReplaceOptions;
 
 import com.couchbase.client.core.msg.kv.DurabilityLevel;
 import com.couchbase.client.java.kv.PersistTo;
+import com.couchbase.client.java.kv.ReplaceOptions;
 import com.couchbase.client.java.kv.ReplicateTo;
 
+/**
+ * Insert Operations
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ExecutableReplaceByIdOperation.java:32]: fixed.

File: pom.xml [hunk 5]
```diff
@@ -173,6 +173,7 @@
         <dependency>
             <groupId>io.projectreactor</groupId>
             <artifactId>reactor-test</artifactId>
+            <version>3.1.0.RELEASE</version>
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply pom.xml:176]: Updated to 3.4.6.  Seems ok.

File: src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java (hunk 2 above)
Comment (mikereiche) [reply src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java:62]: There are some tests that don't delete everything they've inserted.  So I will remove this change from the changeset and leave it for later when I have had time to verify that all the tests clean up after themselves.

File: src/main/java/org/springframework/data/couchbase/core/support/InScope.java [hunk 6]
```diff
@@ -16,11 +16,16 @@
...
 /**
- * A common interface for all of Insert, Replace, Upsert that take Collection
+ * A common interface for all of Insert, Replace, Upsert that take options
  *
  * @author Michael Reiche
  * @param <T> - the entity class
  */
-public interface WithCollection<T> {
-	Object inCollection(String collectionName);
+public interface InScope<T> {
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/support/InScope.java:24]: Yes, it's not used.   Ideally, the methods in all these With* interfaces would return a parameterized interface as in the Override and implementation.  However, there is no common interface for all the different operations.  So I just define it to return Object, and T is not used.  I would like to leave the parameter there, in hopes that I eventually figure out how to define it properly.

	interface FindByQueryInScope<T> extends FindByQueryInCollection<T>, InScope<T> {
		/**
		 * With a different scope
		 *
		 * @param scope the scope to use.
		 */
		@Override
		FindByQueryInCollection<T> inScope(String scope);
	}

File: src/main/java/org/springframework/data/couchbase/core/support/WithAnalyticsConsistency.java [hunk 7]
```diff
@@ -18,11 +18,16 @@
 import com.couchbase.client.java.analytics.AnalyticsScanConsistency;
 
 /**
- * A common interface for all of Insert, Replace, Upsert that take consistency
+ * Interface for operations that take AnalyticsScanConsistency
  *
  * @author Michael Reiche
  * @param <T> - the entity class
  */
 public interface WithAnalyticsConsistency<T> {
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/support/WithAnalyticsConsistency.java:26]: Yes, it's not used. Ideally, the methods in all these With* interfaces would return a parameterized interface as in the Override and implementation. However, there is no common interface for all the different operations. So I just define it to return Object, and T is not used. I would like to leave the parameter there, in hopes that I eventually figure out how to define it properly.

File: src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperation.java [hunk 8]
```diff
@@ -21,35 +21,102 @@
...
-import org.springframework.data.couchbase.core.support.WithCollection;
+import org.springframework.data.couchbase.core.support.WithReplaceOptions;
 
 import com.couchbase.client.core.msg.kv.DurabilityLevel;
 import com.couchbase.client.java.kv.PersistTo;
+import com.couchbase.client.java.kv.ReplaceOptions;
 import com.couchbase.client.java.kv.ReplicateTo;
 
+/**
+ * Insert Operations
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperation.java:35]: fixed

File: src/main/java/org/springframework/data/couchbase/repository/query/AbstractCouchbaseQuery.java [hunk 9]
```diff
@@ -86,7 +87,8 @@ protected Object doExecute(CouchbaseQueryMethod method, ResultProcessor processo
 		ExecutableFindByQuery<?> find = typeToRead == null ? findOperationWithProjection //
 				: findOperationWithProjection; // not yet implemented in core .as(typeToRead);
 
-		String collection = "_default._default";// method.getEntityInformation().getCollectionName(); // not yet implemented
+		// TODO (maybe) // method.getEntityInformation().getCollectionName(); // not yet implemented
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/repository/query/AbstractCouchbaseQuery.java]: This is irrelevant now.  This _default._default is ignored.

**Review by daschl — APPROVED** [review]

**Pull request merged**